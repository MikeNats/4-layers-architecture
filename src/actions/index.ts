import { ActionsStatusType } from './types'
import { requestTransactions, responseTransactions, failGetTransactions } from '../components/trancations/actions'
export {  responseTransactions, requestTransactions } 


export const asyncActions: ActionsStatusType = {
    'TRANSACTIONS': {
      'REQUEST': requestTransactions,
      'SUCCESS': responseTransactions, 
      'FAIL': failGetTransactions
    }
  }
   