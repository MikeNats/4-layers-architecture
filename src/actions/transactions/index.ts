import { Dispatch } from 'redux'
import { 
  RequestTransactionsActionType,
  TransactionPayLoadType, 
  ResponseTransactionsActionType,
  FailGetTransactionsActionType,
  TransactionsActionTypes,
  ACTIONS_ENUM_TYPES
} from './types';
import axios from 'axios';

export const requestTransactions = (index: number): RequestTransactionsActionType => ({
  type: ACTIONS_ENUM_TYPES.REQUEST_TRANSACTIONS,
  index
})

export const responseTransactions = (payload: TransactionPayLoadType): ResponseTransactionsActionType => ({
  type: ACTIONS_ENUM_TYPES.RECEIVE_TRANSACTIONS,
  payload
})

export const failGetTransactions = (errorCode: number): FailGetTransactionsActionType => ({
  type: ACTIONS_ENUM_TYPES.FAILED_TRANSACTIONS,
  errorCode
})

export const fetchtransactions = (n: number): Function =>
  (dispatch: Dispatch): Promise<TransactionsActionTypes> =>  {
    dispatch(requestTransactions(n));

    return axios.get(`http://localhost:3000/mockData/transactions.json`)
      .then(response => dispatch(responseTransactions(response.data)))
      .catch(error => dispatch(failGetTransactions(error.errorCode)))
  }

