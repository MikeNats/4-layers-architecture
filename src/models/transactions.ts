export interface TransactionItemType{
  date: string;
  amount: number;
  product: string;
  image: string
}
 