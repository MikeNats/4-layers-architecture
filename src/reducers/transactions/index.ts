import {
  ACTIONS_ENUM_TYPES,
  TransactionsActionTypes,
} from "../../actions/transactions/types";
import { TransactionsIitialStateType } from './types'

export const initialState: TransactionsIitialStateType = {
  transactions: {
    isFetching: false,
    didInvalidate: false,
    payload: [],
    errorCode: undefined
  }
}

export const transactions = (state: TransactionsIitialStateType = initialState, action: TransactionsActionTypes): TransactionsIitialStateType  => {
  switch (action.type) {
    case ACTIONS_ENUM_TYPES.REQUEST_TRANSACTIONS:
      return {
        ...state,
        transactions: {
          ...state.transactions,
          isFetching: true,
          didInvalidate: false,
          errorCode: undefined
        }
      }
    case ACTIONS_ENUM_TYPES.RECEIVE_TRANSACTIONS:
      return {
        ...state,
        transactions: {
          isFetching: false,  
          didInvalidate: false,
          errorCode: undefined,
          payload: action.payload,
        }
      }
    case ACTIONS_ENUM_TYPES.FAILED_TRANSACTIONS:
      return {
        ...state,
        transactions: {
          ...state.transactions,
          isFetching: false,
          didInvalidate: false,
          errorCode:  action.errorCode,
        }
      }
    default:
      return state
    }
  }