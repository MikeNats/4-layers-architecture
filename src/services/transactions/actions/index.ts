import { Dispatch } from 'redux'
import { 
  TransactionPayloadItemType, 
  TransactionsActionTypes,
  RequestTransactionsActionType,
  ResponseTransactionsActionsType,
  FailGetTransactionsActionType,
  ACTIONS_ENUM_TYPES
} from './types';
import axios from 'axios';

export const requestTransactions = (): RequestTransactionsActionType => ({
  type: ACTIONS_ENUM_TYPES.REQUEST_TRANSACTIONS
})

export const responseTransactions = (payload:TransactionPayloadItemType[]): ResponseTransactionsActionsType => ({
  type: ACTIONS_ENUM_TYPES.RECEIVE_TRANSACTIONS,
  payload
})

export const failGetTransactions = (errorCode: number): FailGetTransactionsActionType => ({
  type: ACTIONS_ENUM_TYPES.FAILED_TRANSACTIONS,
  errorCode
})

export const fetchtransactions = (): Function => 
  (dispatch: Dispatch<TransactionsActionTypes>): Promise<TransactionsActionTypes> =>  {
    dispatch(requestTransactions());

    return axios.get(`http://localhost:3000/mockData/transactions.json`)
      .then(response => dispatch(responseTransactions(response.data)))
      .catch(error => dispatch(failGetTransactions(error.errorCode)))
  }