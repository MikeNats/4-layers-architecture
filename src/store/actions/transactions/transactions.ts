import { 
  TransactionItemType, 
  RequestTransactionsActionType,
  ResponseTransactionsActionsType,
  FailGetTransactionsActionType,
  ACTIONS_ENUM_TYPES
} from './types'; 

export const requestTransactions = (): RequestTransactionsActionType => ({
  type: ACTIONS_ENUM_TYPES.REQUEST_TRANSACTIONS
})

export const responseTransactions = (payload:TransactionItemType[]): ResponseTransactionsActionsType => ({
  type: ACTIONS_ENUM_TYPES.RECEIVE_TRANSACTIONS,
  payload
})

export const failGetTransactions = (errorCode: number): FailGetTransactionsActionType => ({
  type: ACTIONS_ENUM_TYPES.FAILED_TRANSACTIONS,
  errorCode
})