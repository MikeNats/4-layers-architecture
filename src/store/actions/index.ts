import { ActionsStatusType } from './types'
import { requestTransactions, responseTransactions, failGetTransactions } from './transactions/transactions'
export {  responseTransactions, requestTransactions } 


export const asyncActions: ActionsStatusType = {
  'TRANSACTIONS': {
    'REQUEST': requestTransactions,
    'SUCCESS': responseTransactions, 
    'FAIL': failGetTransactions
  }
}
   