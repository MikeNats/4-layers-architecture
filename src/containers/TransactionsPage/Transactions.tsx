import React from 'react';
import { throttle } from 'lodash'
import { connect } from 'react-redux'
import { asyncActions } from '../../store/actions'
import { Dispatch } from 'redux'
import fetch from '../../service/fetch'
import { AsyncActionNames } from '../../store/actions/types'
import { HomePropsType, StateType, HomeState, SortingType } from './types'
import TransactionSearchForm from '../../components/Transactions/TransactionsSearchList/TransactionsSearchForm'
import TransactionsListItem from '../../components/Transactions/TransactionsListItem/TransactionsListItem'
import { Redirect} from "react-router";
import PATHS from '../../routes/PATHS';
import { shortTransactions } from './utils'


class Transactions extends React.Component <HomePropsType, HomeState> {
  constructor(props: HomePropsType) {
    super(props);//componemt did update 
    this.state = {
      short: null ,
      searchTerm: ''   
    };

    this.sortAndFilter = this.sortAndFilter.bind(this)
    this.setSearchTerm = this.setSearchTerm.bind(this)
    this.setSort = this.setSort.bind(this)
  }

  componentDidMount() {//component did update active page man mh kanei loading state not sucess not loop i9f error
    const { getTransactions } = this.props;
    if (!this.props.errorCode) {
      getTransactions(this.props.context.config.TRANSACTIONS_API_URI, this.props.context.userId)
    }  
  }

  private sortAndFilter(){
    return shortTransactions(this.state.short, this.state.searchTerm ? this.props.transactions.filter(item => 
      item.product.toLowerCase().includes(this.state.searchTerm)) : this.props.transactions)
  }

  private setSort(short: SortingType){
    this.setState({short})
  }

  private setSearchTerm = throttle(term => 
    this.setState({searchTerm: term}), 500)
 
  private transactionSearchList(){
    return (
      <main className={`base-layout`}>
        <section className="flex-grow-1">
          <TransactionSearchForm 
            shortTransactions={this.setSort}
            searchItems={this.setSearchTerm}/>
        </section>
        <section className="flex-grow-6">
          { this.sortAndFilter().map(item =>
              <TransactionsListItem 
                key={item.product} 
                className="flex-direction-row" {...item}/>) 
          }   
        </section>
      </main>)
  }
  private redirectToErrorPage() {
    return <Redirect to={{
      pathname: PATHS.ERROR,
      state: { errorCode : this.props.errorCode}}}/>
  }
render() { 
  if (this.props.errorCode) {
    return  this.redirectToErrorPage()
  }else if (this.props.transactions && this.props.transactions.length > 0) {
    return this.transactionSearchList();
    }
      return null;
    }
}
const mapDispatchToProps = (dispatch: Dispatch) => ({
  getTransactions: (url: string, userId: number) => fetch({
    url: `${url}/${userId}`,
    method: 'get',
  }, {
    asyncActionName: asyncActions[AsyncActionNames.TRANSACTIONS]
  })(dispatch)
})

const mapStateToProps = (state: StateType) => ({//lodash check object
  transactions:state.transactions.payload,
  isFetching: state.transactions.isFetching,
  didInvalidate: state.transactions.didInvalidate,
  errorCode: state.transactions.errorCode,
})

export default connect(mapStateToProps, mapDispatchToProps)(Transactions);