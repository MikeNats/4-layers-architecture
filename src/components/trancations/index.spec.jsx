import React from 'react'
import Transactions from './index'
import { shallow, mount } from 'enzyme';
import thunk from 'redux-thunk';
import Adapter from 'enzyme-adapter-react-16';
import Enzyme from 'enzyme';
import configureStore from 'redux-mock-store';
import TransactionSearchList  from '../transactionsSearchList'
import { fetchtransactions } from '../../services/transactions/actions'
jest.mock('../../services/transactions/actions');

Enzyme.configure({ adapter: new Adapter() })

const initialState = { transactions: {
  payload:[],
  isFetching: false,
  didInvalidate: false,
  errorCode: 0,}
}
const middlewares = [thunk], 
 	mockStore = configureStore(middlewares),
   store = mockStore(initialState);
 
   
  const initializedStored  = mockStore({ transactions: {
    payload:[{ 
      date: '2020-04-23T14:40:55.275Z',
      amount: 12312,
      product: 'string',
      image: 'string'}],
    isFetching: false,
    didInvalidate: false,
    errorCode: 0,}
  })
describe("Transactions", () => {
  describe('component', () => {
    fetchtransactions.mockImplementation(() => () => {});
    it('should invoke fetchtransactions on componentDidMount', () => {	
      const component = shallow(<Transactions store={initializedStored} />);
      expect(component).toMatchSnapshot();
    });
    it('should invoke fetchtransactions on componentDidMount', () => {	
      const component = mount(<Transactions store={initializedStored} />).find(TransactionSearchList);
      expect(component.length).toEqual(1);
    });
    it('should should render TransactionSearchList if props are initiazined', () => {	
      fetchtransactions.mockImplementation(() => () => {});

		  mount(<Transactions store={initializedStored} />).find(TransactionSearchList);
      expect(fetchtransactions).toHaveBeenCalled()
  	});

  });
});
