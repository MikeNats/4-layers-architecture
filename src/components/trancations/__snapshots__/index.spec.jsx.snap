// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Transactions should invoke fetchtransactions on componentDidMount 1`] = `ShallowWrapper {}`;
